### PR TITLE
Add Validation in CreateLesson and UpdateLesson methods

### LINKED ISSUE #257: Fix mentorID, studentGroupID and studentID isExist when updating and creating lessons.

### PR DESCRIPTION
**Pull request by QuestroL**
add validation method in LessonService
add new validetion for create and update  lesson
add GetAllStudentInGroup in StudetGroupRepository

### CODE REVIEW

File: CharlieBackend.Business/Services/LessonService.cs [hunk 1]
```diff
@@ -39,6 +40,35 @@ public async Task<Result<LessonDto>> CreateLessonAsync(CreateLessonDto lessonDto
...
+                }
+
+                var foundStudentGroup = await _unitOfWork.StudentGroupRepository.GetByIdAsync(lessonDto.StudentGroupId);
+
+                if (foundStudentGroup == null)
+                {
+                    return Result<LessonDto>.GetError(ErrorCode.NotFound, "Student group is not found");
+                }
+
+                if (foundStudentGroup.StudentsOfStudentGroups.Count != lessonDto.LessonVisits.Count)
```
Comment (SerjKocherga) [inline CharlieBackend.Business/Services/LessonService.cs:59]: Why this parameter should be equal?
Not all of group members could be at a lesson.

File: CharlieBackend.Business/Services/LessonService.cs [hunk 2]
```diff
@@ -39,6 +40,35 @@ public async Task<Result<LessonDto>> CreateLessonAsync(CreateLessonDto lessonDto
...
+                {
+                    return Result<LessonDto>.GetError(ErrorCode.ValidationError, "Something wrong with a Lesson Visits");
+                }
+
+                if (!await CheckStudentsID(lessonDto.LessonVisits, lessonDto.StudentGroupId))
+                {
+                    return Result<LessonDto>.GetError(ErrorCode.NotFound, "Someone of student is not found or not included in this group");
```
Comment (SerjKocherga) [inline CharlieBackend.Business/Services/LessonService.cs]: Make this message more informative, add name of a student and his id.

File: CharlieBackend.Business/Services/LessonService.cs [hunk 3]
```diff
@@ -39,6 +40,35 @@ public async Task<Result<LessonDto>> CreateLessonAsync(CreateLessonDto lessonDto
...
+                }
+
+                if (createdLessonEntity.LessonDate < DateTime.Now)
+                {
+                    return Result<LessonDto>.GetError(ErrorCode.ValidationError, "Lesson date is incorrect");
```
Comment (SerjKocherga) [inline CharlieBackend.Business/Services/LessonService.cs:73]: Why it is incorrect? Please, provide more corrective information for a user.

File: CharlieBackend.Business/Services/LessonService.cs [hunk 4]
```diff
@@ -168,5 +198,32 @@ public async Task<Result<IList<StudentLessonDto>>> GetStudentLessonsAsync(long s
...
+            }
+
+            var checkStudentInGroup = await _unitOfWork.StudentGroupRepository.GetAllStudentInGroup(groupId);
+
+            if (checkStudentInGroup.Count != studentsId.Count)
+            {
+                return false;
+            }
+
+            foreach (var item in studentsId)
```
Comment (SerjKocherga) [inline CharlieBackend.Business/Services/LessonService.cs:220]: Why do you need this check?
Seems like it always be true.

File: CharlieBackend.Data/Repositories/Impl/StudentGroupRepository.cs [hunk 5]
```diff
@@ -107,6 +107,12 @@ public async Task<List<MentorStudyGroupsDto>> GetMentorStudyGroups(long id)
...
+        }
+
+        public async Task<IList<long?>> GetAllStudentInGroup(long id)
+        {
+            return await _applicationContext.StudentsOfStudentGroups.Where(s => s.StudentGroupId == id)
+                .Select(s => s.StudentId).ToListAsync();
+        }
     }
-}
+}
```
Comment (SerjKocherga) [inline CharlieBackend.Data/Repositories/Impl/StudentGroupRepository.cs:118]: Add a new empty line at the end of file.

File: CharlieBackend.Business/Services/LessonService.cs (hunk 4 above)
Comment (QuestroL) [reply CharlieBackend.Business/Services/LessonService.cs:220]: it is possible that the  mentor has writed wrong StudentId or  student may not be  included in group on this lesson

File: CharlieBackend.Business/Services/LessonService.cs [hunk 6]
```diff
@@ -39,6 +40,35 @@ public async Task<Result<LessonDto>> CreateLessonAsync(CreateLessonDto lessonDto
...
+                var foundStudentGroup = await _unitOfWork.StudentGroupRepository.GetByIdAsync(lessonDto.StudentGroupId);
+
+                if (foundStudentGroup == null)
+                {
+                    return Result<LessonDto>.GetError(ErrorCode.NotFound, "Student group is not found");
+                }
+
+                if (foundStudentGroup.StudentsOfStudentGroups.Count != lessonDto.LessonVisits.Count)
+                {
+                    return Result<LessonDto>.GetError(ErrorCode.ValidationError, "Something wrong with a Lesson Visits");
```
Comment (whereismyroot) [inline CharlieBackend.Business/Services/LessonService.cs]: Describe what is actually wrong, or describe correct condition.

File: CharlieBackend.Business/Services/LessonService.cs [hunk 7]
```diff
@@ -168,5 +198,32 @@ public async Task<Result<IList<StudentLessonDto>>> GetStudentLessonsAsync(long s
...
         }
 
+        private async Task<bool> CheckStudentsID(IList<VisitDto> visit, long groupId)
+        {
+            List<long> studentsId = new List<long>();
+
+            foreach (var item in visit)
+            {
+                studentsId.Add((long)item.StudentId);
+            }
```
Comment (whereismyroot) [inline CharlieBackend.Business/Services/LessonService.cs]: var studentsId = visit.Select(x => x.StudentId.Value).ToList();

File: CharlieBackend.Data/Repositories/Impl/Interfaces/IStudentGroupRepository.cs [hunk 8]
```diff
@@ -14,6 +14,8 @@ public interface IStudentGroupRepository : IRepository<StudentGroup>
 
         Task<List<MentorStudyGroupsDto>> GetMentorStudyGroups(long id);
 
+        Task<IList<long?>> GetAllStudentInGroup(long id);
```
Comment (whereismyroot) [inline CharlieBackend.Data/Repositories/Impl/Interfaces/IStudentGroupRepository.cs]: GetGroupStudentsIds

File: CharlieBackend.Business/Services/LessonService.cs (hunk 1 above)
Comment (whereismyroot) [reply CharlieBackend.Business/Services/LessonService.cs:59]: That's correct, but for each student in group should be created visit, either with Presence= 0 or 1

**Review by whereismyroot — CHANGES_REQUESTED** [review]
small changes

**Review by filinskyi — APPROVED** [review]

File: CharlieBackend.Data/Repositories/Impl/StudentGroupRepository.cs [hunk 9]
```diff
@@ -1,4 +1,5 @@
 ﻿using CharlieBackend.Core.DTO.Mentor;
+using CharlieBackend.Core.DTO.Mentor;
```
Comment (SnipeFrost) [inline CharlieBackend.Data/Repositories/Impl/StudentGroupRepository.cs]: two same using

**Review by SnipeFrost — APPROVED** [review]

**Review by Hlib-mats — APPROVED** [review]

**Review by whereismyroot — APPROVED** [review]

**Pull request merged**